### PR TITLE
Add --validValue option to prop print command

### PR DESCRIPTION
**Pull request by rosen-vladimirov**
Add validValue to options. When it is specified, all valid values of the requested property will be printed. Improve prop print command to work in the following way:
I. inside project dir:
     1. $ appbuilder prop print -> prints all properties of current project and their current values
     2. $ appbuilder prop print --validValue -> prints all properties valid for the current project and all valid values for each of them.
     3. $ appbuilder prop print <PropName> -> prints the value of the property for the current project
     4. $ appbuilder prop print <PropName> --validValue -> prints all valid values for the selected property, applicable to the current project type.

II. Outside of project dir:
     1. $ appbuilder prop print -> leads to error, that no project is found
     2. $ appbuilder prop print --validValue -> prints all properties for all projects and all valid values for each of them.
     3. $ appbuilder prop print <PropName> -> leads to error, that no project is found
     4. $ appbuilder prop print <PropName> --validValue -> prints all valid values for the selected property

Remove dynamic call from help.txt in order to make the output more readable. Remove calls to ensureProject when working with prop print command.

http://teampulse.telerik.com/view#item/284655

### CODE REVIEW

File: lib/project/project.d.ts [hunk 1]
```diff
@@ -44,8 +44,6 @@ declare module Project {
 		requiredAndroidApiLevel: number;
 		configFiles: IConfigurationFile[];
 		getTemplateFilename(name: string): string;
-		projectTemplatesString(): IFuture<string>;
-		alterPropertiesForNewProject(properties: any, projectName: string): void;
```
Comment (rosen-vladimirov) [inline lib/project/project.d.ts]: Removed as they are duplicated :)

**Comment by rosen-vladimirov** [comment]
ping @ikoevska  for the help.txt - we have to improve the information about prop print command. I've written just the base idea, but maybe we should include more information how the command works in different situatuions (check PR's description or teampulse item for more info).

File: lib/json-schema/json-schema-validator.ts [hunk 2]
```diff
@@ -63,18 +63,19 @@ export class JsonSchemaValidator implements IJsonSchemaValidator {
...
-				});
-			}
 		}
 
+        var projectPropertiesFilePath = this.$resources.resolvePath(util.format("project-properties-%s.json",framework.toLowerCase()));
+        if(this.$fs.exists(projectPropertiesFilePath).wait()) {
+            var fileContent = this.$fs.readJson(projectPropertiesFilePath).wait();
+            var additionalProperties = _.keys(fileContent);
+            _.each(additionalProperties, (propertyName: string) => {
+                _.each(_.keys(fileContent[propertyName]), (value: string) => {
```
Comment (teobugslayer) [inline lib/json-schema/json-schema-validator.ts]: We read the keys, but the callback receives values? Which is which?

File: lib/declarations.d.ts [hunk 3]
```diff
@@ -199,6 +199,14 @@ interface IProjectPropertiesService {
 	completeProjectProperties(properties: any, frameworkProject: Project.IFrameworkProject): boolean;
 	updateProjectProperty(projectData: any, mode: string, property: string, newValue: any): IFuture<void>;
 	normalizePropertyName(property: string, projectData: IProjectData): string;
+	getValidValuesForProperty(propData: any): IFuture<any[]>
+	getPropertiesForAllSupportedProjects(): IFuture<string>;
+}
+
+interface IPropertyValue {
+	id: string;
```
Comment (teobugslayer) [inline lib/declarations.d.ts]: Is this member really needed?

File: lib/options.ts [hunk 4]
```diff
@@ -22,7 +22,8 @@ var knownOpts: any = {
 		"client": String,
 		"available": Boolean,
 		"release": Boolean,
-		"debug": Boolean
+		"debug": Boolean,
+		"validValue": Boolean
```
Comment (teobugslayer) [inline lib/options.ts]: will $ appbuilder blah --validvalue (note the lower case) work on Windows?

File: lib/project.ts [hunk 5]
```diff
@@ -358,23 +359,77 @@ export class Project implements Project.IProject {
...
+						_.each(_.keys(schema), propKey => {
+							var prop = schema[propKey];
+							this.$logger.info("  " + propKey);
+							this.printValidValuesOfProperty(prop).wait();
+						});
+					} else {
+						// 'appbuilder prop print' called inside project dir
+						var propKeys: any = _.keys(this.projectData);
+						var sortedProperties = _.sortBy(propKeys, (propertyName: string) => {
+							return propertyName.toLowerCase();
```
Comment (teobugslayer) [inline lib/project.ts]: Maybe we can move toLowerCase outside sorting? for increased efficiency ;)

File: lib/project.ts [hunk 6]
```diff
@@ -358,23 +359,77 @@ export class Project implements Project.IProject {
...
+						if(_.has(this.projectData, normalizedPropertyName)) {
+							this.$logger.out(this.projectData[normalizedPropertyName]);
+						} else {
+							this.$errors.fail("Unrecognized project property '%s'", property);
+						}
+					}
+				} else {
+					if(options.validValue) {
+						// 'appbuilder prop print --validValue' called inside project dir
+						_.each(_.keys(schema), propKey => {
```
Comment (rosen-vladimirov) [inline lib/project.ts]: sort the keys

File: lib/options.ts (hunk 4 above)
Comment (rosen-vladimirov) [inline lib/options.ts]: no, we have two options here: add valid-value as known option or change validValue to be only valid.
Ping @Fatme @ikoevska  - which one do you think is better?

File: lib/project.ts [hunk 7]
```diff
@@ -358,23 +359,77 @@ export class Project implements Project.IProject {
...
+			if(property.pattern) {
+				this.$logger.trace("%sDesired pattern is: %s", Project.INDENTATION, property.pattern);
 			}
 
+			var validValues: IPropertyValue[] = this.$projectPropertiesService.getValidValuesForProperty(property).wait();
+			if(validValues && validValues.length > 0) {
+				this.$logger.out("%sValid values:", Project.INDENTATION);
+			}
+			_.forEach(validValues, (val: any) => {
+				this.$logger.out("%s  %s    %s", Project.INDENTATION, val.id, val.description);
```
Comment (teobugslayer) [inline lib/project.ts]: We can require("table") here and pretty-print it.

File: lib/services/project-properties-service.ts [hunk 8]
```diff
@@ -162,6 +168,21 @@ export class ProjectPropertiesService implements IProjectPropertiesService {
 		}).future<string[]>()();
 	}
 
+	public getValidValuesForProperty(propData: any): IFuture<IPropertyValue[]> {
+		return ((): IPropertyValue[]=> {
+			var validValues: PropertyValue[] = [];
+			var range = this.getPropRange(propData).wait();
+			if(range) {
+				_.each(range, (rangeDesc: any, rangeKey: any) => {
```
Comment (teobugslayer) [inline lib/services/project-properties-service.ts]: Why not _.map range -> validValues?

File: lib/project.ts (hunk 7 above)
Comment (Fatme) [inline lib/project.ts]: It seems that somethimes `val.description` is `undefined` and we print `undefined` on the console. I got the following output 
`Framework
     Valid values:
       Cordova    undefined` when I execute `appbuilder prop print --validValue`. I suggest to print the `description`  only when we have such.

File: lib/project.ts (hunk 5 above)
Comment (rosen-vladimirov) [inline lib/project.ts]: Well, the idea is to persist the prioperties as they are, but to sort them based on their toLowerCase values. That's why toLowerCase is inside the sortBy

File: lib/project.ts [hunk 9]
```diff
@@ -363,23 +364,82 @@ export class Project implements Project.IProject {
...
-				this.$errors.fail("Unrecognized project property '%s'", property);
+			if(property.pattern) {
+				this.$logger.trace("%sDesired pattern is: %s", Project.INDENTATION, property.pattern);
 			}
 
+			var validValues: string[] = this.$projectPropertiesService.getValidValuesForProperty(property).wait();
+			if(validValues) {
+				this.$logger.out("%sValid values:", Project.INDENTATION);
+			}
+			_.forEach(validValues, value => {
```
Comment (rosen-vladimirov) [inline lib/project.ts]: maybe move the _.forEach inside the above if

**Comment by Fatme** [comment]
:+1:

**Comment by teobugslayer** [comment]
After discussing in person, :+1:

**Pull request merged**